salt/state.py

```diff
@@ -1261,6 +1294,15 @@ def compile_high_data(self, high):
                                   not isinstance(val, six.string_types)):
                                 # Invalid name, fall back to ID
                                 chunk[key] = name
+                            elif ('__sls__' in body and
+                                  '__env__' in body and
+                                  key in STATE_REQUISITE_KEYWORDS):
+                                _val = type(val)()
```

Comment:
Would it be safer to stick with an `OrderedDict`, since we wouldn't be losing data?